spanner/google/cloud/spanner_v1/snapshot.py

```diff
@@ -105,7 +105,7 @@ def read(self, table, columns, keyset, index='', limit=0, partition=None):
 
         :type partition: bytes
         :param partition: (Optional) one of the partition tokens returned
-                          from :meth:`partition_read`.  Incompatible with
+                          from :meth:`_partition_read`.  Incompatible with
```

Comment:
I think we should remove this parameter from this method and instead create a private overloaded method that takes this parameter as well.

Reply:
Python doesn't have the concept of "overloads". 

Do we expect users to call `read` with `partition` or are we expecting just our helpers to do so? If the former, we should leave as-is, if the later, we can definitely move this to a private `_read_from_partition` (which is unfortunately named because we also have `_partition_read`.

Reply:
We expect users to use the methods in BatchSnapshot to read/query from partitions and not directly invoke the method in Snapshot. Also note that the way this is documented right now, it tells users that to pass this parameter, they need to invoke a private method `_partition_read`.
I am ok with whatever name you chose for the private method.

Reply:
Sounds good. There should be preferably one obvious way to do things. @tseaver, can you handle this?

Reply:
@jonparrott  I would rather back out fad5366:  ISTM that the `Snapshot.partition_{read,query}` API wrapper methods are exactly like `Session.{create,delete}`:  they are lower-level methods which most users won't call directly, but we have no reason to prevent them from doing so.

Reply:
Okay, so help me understand here. We should have one (preferably obvious) way to accomplish starting a partition read/query and then reading data from it.

Can we pick whichever way that is and make everything else either private or remove it? What's stopping us from doing this?

Reply:
I think there is a good reason to prevent people from directly using the snapshot interface for partitioned reads. If they use the snapshot interface how will they make sure that they are using the same session id and transaction id across machines?

Reply:
Here is the "obvious" / documented / expected way:

```python
        batch_txn = self._db.batch_snapshot(read_timestamp=committed)
        for batch in batch_txn.generate_query_batches(sql):
            do_something_with_rows( batch_txn.process(batch))  # might  be on another host
        batch_txn.close()
```

`Database.generate_query_batches` is a convenience wrapper:  it calls the underling `Snapshot.partition_query` method and returns results suitable for serialization / processing by `Database.process()`.

Making the lower level `partition_query` private is a bit BDSM for Python:  it is the wrapper we have for the published `PartitionQueryRequest` API.  In this case, it is like the `Session.create` / `Session.delete` API wrappers:  they are low-level API wrappers used by the higher-level database convenience methods.

Reply:
Gotcha. Your position makes sense to me. Let's do that and get this merged.

Reply:
Rebased with fad5366 backed out.